fs/smb/client/cifsacl.c

```diff
@@ -1809,3 +1819,1126 @@ int cifs_set_acl(struct mnt_idmap *idmap, struct dentry *dentry,
+	return ACES_TO_XDRSIZE(1);
+}
+
+/*
+ * Per Microsoft Win32 documentation, a NULL DACL grants full access to
+ * any user that requests it; normal security checking is not performed
+ * with respect to the object. We can't set a NULL ZFS ACL and so
+ * the best we can do is set one granting everyone@ full control.
+ */
+static int
+generate_null_zfsacl(char **buf_out)
+{
+	u32 *xdrbuf = NULL, *zfsacl;
+
+	xdrbuf = kzalloc(ACES_TO_XDRSIZE(1), GFP_KERNEL);
+	if (!xdrbuf)
+		return -ENOMEM;
+
+	zfsacl = xdrbuf;
+	*zfsacl++ = 0; /* acl_flags */
+	*zfsacl++ = htonl(1); /* acl count */
+
+	set_xdr_ace(zfsacl, ACEI4_SPECIAL_WHO, ACE4_SPECIAL_EVERYONE,
+		    ACE4_ACCESS_ALLOWED_ACE_TYPE, ACE4_ALL_PERMS, 0);
+
+	*buf_out = (char *)xdrbuf;
+	return ACES_TO_XDRSIZE(1);
+}
+
+/*
+ * Convert generic access into NFSv4 perms
+ */
+static u32
+generic_to_nfs(u32 generic_access)
+{
+	u32 out = 0;
+	if (generic_access == GENERIC_ALL) {
+		return ACE4_ALL_PERMS;
+	}
+
+	if (generic_access & GENERIC_READ) {
+		out |= ACE4_READ_PERMS;
+	}
+
+	if (generic_access & GENERIC_EXECUTE) {
+		out |= ACE4_EXECUTE;
+	}
+
+	if (generic_access & GENERIC_WRITE) {
+		out |= (ACE4_WRITE_PERMS|ACE4_DELETE);
+	}
+
+	return out;
+}
+
+static int
+convert_smb_access_to_nfs(u32 smbaccess, u32 *nfs_access_out)
+{
+	int i;
+	u32 perms = generic_to_nfs(smbaccess);
+
+	if (perms == ACE4_ALL_PERMS) {
+		*nfs_access_out = perms;
+		return 0;
+	}
+
+	for (i = 0; i < (sizeof(nfsperm2smb) / sizeof(nfsperm2smb[0])); i++) {
+		if (smbaccess & nfsperm2smb[i].smb_perm)
+			perms |= nfsperm2smb[i].nfs_perm;
+	}
+
+	*nfs_access_out = perms;
+	return 0;
+}
+
+static int
+convert_smb_flags_to_nfs(u8 smbflags, u32 *nfs_flags_out)
+{
+	int i;
+	u32 flags = 0;
+
+	if (smbflags & (SUCCESSFUL_ACCESS_ACE_FLAG | FAILED_ACCESS_ACE_FLAG)) {
+		cifs_dbg(VFS, "%s: ACE contains unsupported flags 0x%04x\n",
+			 __func__, smbflags);
+		return -EINVAL;
+	}
+
+	for (i = 0; i < (sizeof(nfsflag2smb) / sizeof(nfsflag2smb[0])); i++) {
+		if (smbflags & nfsflag2smb[i].smb_flag)
+			flags |= nfsflag2smb[i].nfs_flag;
+	}
+
+	*nfs_flags_out = flags;
+	return 0;
+}
+
+static int
+convert_smb_ace_type_to_nfs(u8 smbacetype, u32 *nfs_ace_type_out)
+{
+	switch (smbacetype) {
+	case ACCESS_ALLOWED_ACE_TYPE:
+		*nfs_ace_type_out = ACE4_ACCESS_ALLOWED_ACE_TYPE;
+		break;
+	case ACCESS_DENIED_ACE_TYPE:
+		*nfs_ace_type_out = ACE4_ACCESS_DENIED_ACE_TYPE;
+		break;
+	default:
+		cifs_dbg(VFS, "%s: ACE contains unsupported ace type 0x%04x\n",
+			 __func__, smbacetype);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/*
+ * Convert SID into NFS4 ID and type. Try to map BUILTIN / SPECIAL sids
+ * directly to NFS4 special type where possible (to avoid upcall to winbindd).
+ *
+ * Since the existing idmap key implementation does support return of both
+ * users and groups in one call, we first try to retrieve group (because this
+ * is in real life much more likely). If retrieving as group fails, we retry
+ * as user.
+ */
```

Comment:
Maybe add a comment about the return value.

It appears that we have 3 states:

- `< 0`: error
- `> 0`: success
- `== 0`: upcall required

Is that correct?